### PR TITLE
add drop, desc, show index's validator and executor

### PR DESCRIPTION
**Pull request by jievince**
show tag/edge index status will be done in another pr

### CODE REVIEW

**Comment by jievince** [comment]
I will add tests in another pr

File: src/util/SchemaUtil.cpp [hunk 1]
```diff
@@ -290,6 +290,51 @@ StatusOr<DataSet> SchemaUtil::toShowCreateSchema(bool isTag,
...
+    }
+    return dataSet;
+}
+
+StatusOr<DataSet> SchemaUtil::toShowCreateIndex(bool isTagIndex,
+                                                const std::string &indexName,
+                                                const meta::cpp2::IndexItem &indexItem) {
+    DataSet dataSet;
+    std::string createStr;
+    createStr.resize(1024);
```
Comment (laura-ding) [inline src/util/SchemaUtil.cpp:313]: reserve

File: src/executor/maintain/TagIndexExecutor.cpp [hunk 2]
```diff
@@ -20,7 +23,7 @@ folly::Future<Status> CreateTagIndexExecutor::execute() {
             .then([ctiNode, spaceId](StatusOr<IndexID> resp) {
                 if (!resp.ok()) {
                     LOG(ERROR) << "SpaceId: " << spaceId
-                               << ", Create index`" << ctiNode->getIndexName()
+                               << ", Create index `" << ctiNode->getIndexName()
```
Comment (bright-starry-sky) [inline src/executor/maintain/TagIndexExecutor.cpp:26]: A pair of single quotes

File: src/util/SchemaUtil.cpp [hunk 3]
```diff
@@ -290,6 +290,51 @@ StatusOr<DataSet> SchemaUtil::toShowCreateSchema(bool isTag,
...
+    for (auto &col : indexItem.get_fields()) {
+        Row row;
+        row.values.emplace_back(Value(col.get_name()));
+        row.values.emplace_back(typeToString(col));
+        auto nullable = col.__isset.nullable ? *col.get_nullable() : false;
+        row.values.emplace_back(nullable ? "YES" : "NO");
+        auto defaultValue = col.__isset.default_value ? *col.get_default_value() : Value();
+        row.values.emplace_back(std::move(defaultValue));
+        dataSet.emplace_back(std::move(row));
+    }
```
Comment (bright-starry-sky) [inline src/util/SchemaUtil.cpp:304]: if the field type is fixed_string , do we need to handle the type length?

**Review by bright-starry-sky — COMMENTED** [review]
Nice job.  could you add some python test cases? thanks.

File: src/util/SchemaUtil.cpp [hunk 4]
```diff
@@ -290,6 +290,47 @@ StatusOr<DataSet> SchemaUtil::toShowCreateSchema(bool isTag,
...
+    for (auto &col : indexItem.get_fields()) {
+        Row row;
+        row.values.emplace_back(Value(col.get_name()));
+        row.values.emplace_back(typeToString(col));
+        dataSet.emplace_back(std::move(row));
+    }
+    return dataSet;
+}
+
+StatusOr<DataSet> SchemaUtil::toShowCreateIndex(bool isTagIndex,
```
Comment (darionyaphet) [inline src/util/SchemaUtil.cpp:304]: why don't move to `IndexUtil` ?

Comment (jievince) [reply src/util/SchemaUtil.cpp]: Sorry, I don't notice IndexUtil before. I have fixed it just now.

File: src/planner/Maintain.h [hunk 5]
```diff
@@ -573,47 +573,62 @@ class DropEdgeIndex final : public DropIndexNode {
...
+                                    PlanNode* input,
+                                    std::string indexName) {
+        return qctx->objPool()->add(new ShowCreateTagIndex(qctx->genId(),
+                                    input,
+                                    std::move(indexName)));
     }
 
-    std::unique_ptr<cpp2::PlanNodeDescription> explain() const override;
+private:
+    ShowCreateTagIndex(int64_t id,
```
Comment (yixinglu) [inline src/planner/Maintain.h:587]: Now the plan node construction parameter has been changed, you should pass the QueryContext pointer in.

File: src/planner/Maintain.h [hunk 6]
```diff
@@ -573,47 +573,62 @@ class DropEdgeIndex final : public DropIndexNode {
...
+    static ShowCreateEdgeIndex* make(QueryContext* qctx,
+                                     PlanNode* input,
+                                     std::string indexName) {
+        return qctx->objPool()->add(new ShowCreateEdgeIndex(qctx->genId(),
+                                                            input,
+                                                            std::move(indexName)));
+    }
+
+private:
+    ShowCreateEdgeIndex(int64_t id,
```
Comment (yixinglu) [inline src/planner/Maintain.h:605]: ditto

File: src/util/IndexUtil.cpp [hunk 7]
```diff
@@ -22,6 +22,47 @@ Status IndexUtil::validateColumns(const std::vector<std::string>& fields) {
...
+        row.values.emplace_back(SchemaUtil::typeToString(col));
+        dataSet.emplace_back(std::move(row));
+    }
+    return dataSet;
+}
+
+StatusOr<DataSet> IndexUtil::toShowCreateIndex(bool isTagIndex,
+                                                const std::string &indexName,
+                                                const meta::cpp2::IndexItem &indexItem) {
+    DataSet dataSet;
```
Comment (yixinglu) [inline src/util/IndexUtil.cpp:39]: indent

File: src/planner/Maintain.h (hunk 5 above)
Comment (jievince) [reply src/planner/Maintain.h]: Thanks. Has fixed.

File: src/executor/maintain/TagIndexExecutor.cpp [hunk 8]
```diff
@@ -6,40 +6,196 @@
...
+            auto tagIndexItems = std::move(resp).value();
+            nebula::DataSet v({"Name", "Tag Index Status"});
+            for (auto &tagIndex : tagIndexItems) {
+                indexesStatus_.emplace(tagIndex.get_index_name(), "SUCCEEDED");
+            }
+            for (const auto &indexStatus : indexesStatus_) {
+                v.emplace_back(nebula::Row({indexStatus.first, indexStatus.second}));
+            }
+            return finish(std::move(v));
+        });
```
Comment (laura-ding) [inline src/executor/maintain/TagIndexExecutor.cpp:198]: What's meaning?

File: src/executor/maintain/TagIndexExecutor.cpp [hunk 9]
```diff
@@ -6,40 +6,196 @@
...
+                for (const auto &jobDesc : jobsDesc) {
+                    if (jobDesc.get_paras()[0] == spaceName &&
+                        jobDesc.get_cmd() == meta::cpp2::AdminCmd::REBUILD_TAG_INDEX) {
+                        indexesStatus_.emplace(
+                            jobDesc.get_paras()[1],
+                            meta::cpp2::_JobStatus_VALUES_TO_NAMES.at(jobDesc.get_status()));
+                    }
+                }
+                return Status::OK();
+            });
```
Comment (laura-ding) [inline src/executor/maintain/TagIndexExecutor.cpp:179]: Your result not use. and one executor can call async once.

File: src/executor/maintain/EdgeIndexExecutor.cpp [hunk 10]
```diff
@@ -30,21 +33,171 @@ folly::Future<Status> CreateEdgeIndexExecutor::execute() {
...
 
+folly::Future<Status> ShowEdgeIndexStatusExecutor::execute() {
+    SCOPED_TIMER(&execTime_);
+
+    auto spaceInfo = qctx()->rctx()->session()->space();
+    auto spaceName = spaceInfo.name;
+    auto spaceId = spaceInfo.id;
+    auto status = qctx()
+            ->getMetaClient()
+            ->submitJob(meta::cpp2::AdminJobOp::SHOW_All,
```
Comment (laura-ding) [inline src/executor/maintain/EdgeIndexExecutor.cpp:154]: ditto

And I think is repeat with the Job executor

File: src/executor/maintain/EdgeIndexExecutor.cpp [hunk 11]
```diff
@@ -6,45 +6,179 @@
...
+folly::Future<Status> ShowEdgeIndexStatusExecutor::execute() {
+    SCOPED_TIMER(&execTime_);
+
+    auto spaceInfo = qctx()->rctx()->session()->space();
+    auto spaceName = spaceInfo.name;
+    auto spaceId = spaceInfo.id;
+    return qctx()
+        ->getMetaClient()
+        ->submitJob(meta::cpp2::AdminJobOp::SHOW_All,
+                    meta::cpp2::AdminCmd::COMPACT,
```
Comment (laura-ding) [inline src/executor/maintain/EdgeIndexExecutor.cpp:153]: why the cmd is `COMPACT `, modify to `REBUILD_EDGE_INDEX`

File: src/executor/maintain/EdgeIndexExecutor.cpp [hunk 12]
```diff
@@ -6,45 +6,179 @@
...
+                           << resp.status();
+                return resp.status();
+            }
+
+            DCHECK(resp.value().__isset.job_desc);
+            if (!resp.value().__isset.job_desc) {
+                return Status::Error("Response unexpected");
+            }
+
+            nebula::DataSet v({"Name", "Tag Index Status"});
```
Comment (laura-ding) [inline src/executor/maintain/EdgeIndexExecutor.cpp:170]: Edge index status

File: src/executor/maintain/EdgeIndexExecutor.cpp [hunk 13]
```diff
@@ -6,45 +6,179 @@
...
+            const auto &jobsDesc = *resp.value().get_job_desc();
+            for (const auto &jobDesc : jobsDesc) {
+                if (jobDesc.get_paras()[0] == spaceName &&
```
Comment (laura-ding) [inline src/executor/maintain/EdgeIndexExecutor.cpp:173]: check the get_paras()'s size

File: src/executor/Executor.cpp [hunk 14]
```diff
@@ -266,12 +266,30 @@ Executor *Executor::makeExecutor(QueryContext *qctx, const PlanNode *node) {
...
+        case PlanNode::Kind::kShowEdgeIndexes: {
+            return pool->add(new ShowEdgeIndexesExecutor(node, qctx));
+        }
+        case PlanNode::Kind::kShowTagIndexStatus: {
+            return pool->add(new ShowTagIndexStatusExecutor(node, qctx));
+        }
+        case PlanNode::Kind::kShowEdgeIndexStatus: {
+            return pool->add(new ShowEdgeIndexStatusExecutor(node, qctx));
+        }
+        // case PlanNode::Kind::kRebuildTagIndex: {
```
Comment (laura-ding) [inline src/executor/Executor.cpp:287]: delete the unuse code

File: src/executor/maintain/TagIndexExecutor.cpp [hunk 15]
```diff
@@ -6,40 +6,179 @@
...
+folly::Future<Status> ShowTagIndexStatusExecutor::execute() {
+    SCOPED_TIMER(&execTime_);
+
+    auto spaceInfo = qctx()->rctx()->session()->space();
+    auto spaceName = spaceInfo.name;
+    auto spaceId = spaceInfo.id;
+    return qctx()
+        ->getMetaClient()
+        ->submitJob(meta::cpp2::AdminJobOp::SHOW_All,
+                    meta::cpp2::AdminCmd::COMPACT,
```
Comment (laura-ding) [inline src/executor/maintain/TagIndexExecutor.cpp:153]: ditto

File: src/executor/maintain/TagIndexExecutor.cpp [hunk 16]
```diff
@@ -6,40 +6,179 @@
...
+
+            DCHECK(resp.value().__isset.job_desc);
+            if (!resp.value().__isset.job_desc) {
+                return Status::Error("Response unexpected");
+            }
+
+            nebula::DataSet v({"Name", "Tag Index Status"});
+            const auto &jobsDesc = *resp.value().get_job_desc();
+            for (const auto &jobDesc : jobsDesc) {
+                if (jobDesc.get_paras()[0] == spaceName &&
```
Comment (laura-ding) [inline src/executor/maintain/TagIndexExecutor.cpp:173]: ditto, check the size of get_paras

File: src/planner/Maintain.h [hunk 17]
```diff
@@ -573,49 +514,115 @@ class DropEdgeIndex final : public DropIndexNode {
...
-    RebuildEdgeIndex(QueryContext* qctx, PlanNode* input, std::string indexName)
-        : RebuildIndexNode(qctx, Kind::kRebuildEdgeIndex, input, std::move(indexName)) {}
+    ShowEdgeIndexStatus(QueryContext* qctx, PlanNode* input)
+        : SingleInputNode(qctx, Kind::kShowEdgeIndexStatus, input) {}
 };
 
-}  // namespace graph
-}  // namespace nebula
-#endif  // PLANNER_MAINTAIN_H_
+// class RebuildIndexNode : public SingleInputNode {
```
Comment (laura-ding) [inline src/planner/Maintain.h:583]: delete the unuse code

File: src/validator/MaintainValidator.h [hunk 18]
```diff
@@ -334,6 +359,9 @@ class RebuildTagIndexValidator final : public Validator {
     Status validateImpl() override;
 
     Status toPlan() override;
+
+private:
+    std::string                      indexName_;
```
Comment (laura-ding) [inline src/validator/MaintainValidator.h:364]: unused it

File: src/validator/MaintainValidator.h [hunk 19]
```diff
@@ -325,6 +328,28 @@ class ShowEdgeIndexesValidator final : public Validator {
...
+    ShowEdgeIndexStatusValidator(Sentence* sentence, QueryContext* context)
+        : Validator(sentence, context) {}
+
+private:
+    Status validateImpl() override;
+
+    Status toPlan() override;
+};
+
 class RebuildTagIndexValidator final : public Validator {
```
Comment (laura-ding) [inline src/validator/MaintainValidator.h:353]: the RebuildTagIndexValidator unused yet.

File: tests/maintain/test_index.py [hunk 20]
```diff
@@ -70,6 +70,92 @@ def test_tag_index(self):
...
+
+        # resp = self.client.execute_query('REBUILD TAG INDEX non_existent_tag_index')
+        # self.check_resp.succeeded(resp) # need to check if index exists in validator in future
+
+        # Show Tag Index Status
+        # resp = self.client.execute_query('SHOW TAG INDEX STATUS')
+        # self.check_resp_succeeded(resp)
+        # self.check_out_of_order_result(resp, [['single_tag_index', 'SUCCEEDED'], ['multi_tag_index', 'SUCCEEDED'], ['disorder_tag_index', 'SUCCEEDED']])
+
+        # Describe Tag Index
```
Comment (laura-ding) [inline tests/maintain/test_index.py:101]: Add test to lookup data after rebuild

File: tests/maintain/test_index.py [hunk 21]
```diff
@@ -105,3 +191,89 @@ def test_edge_index(self):
...
+                  ['col2', 'int64']]
+        self.check_result(resp, expect)
+
+        resp = self.client.execute_query('DESC EDGE INDEX non_existent_edge_index')
+        self.check_resp_failed(resp)
+
+        # Show Create Edge Index
+        resp = self.client.execute_query('SHOW CREATE EDGE INDEX single_edge_index')
+        self.check_resp_succeeded(resp)
+        expect = [['single_edge_index', 'CREATE EDGE INDEX `single_edge_index` ON `edge_1` (\n `col2`\n)']]
```
Comment (laura-ding) [inline tests/maintain/test_index.py:247]: drop the `single_edge_index `, then use the expect result to create `single_edge_index`

File: tests/maintain/test_index.py [hunk 22]
```diff
@@ -105,3 +191,89 @@ def test_edge_index(self):
...
+        resp = self.client.execute_query('SHOW CREATE EDGE INDEX non_existent_edge_index')
+        self.check_resp_failed(resp)
+
+        # Show Edge Indexes
+        resp = self.client.execute_query('SHOW EDGE INDEXES')
+        self.check_resp_succeeded(resp)
+        self.check_out_of_order_result(resp, [['single_edge_index'], ['multi_edge_1_index'], ['disorder_edge_1_index']])
+
+        # Drop Edge Index
+        resp = self.client.execute_query('DROP EDGE INDEX single_edge_index')
```
Comment (laura-ding) [inline tests/maintain/test_index.py:269]: need to check the index is drop, such as check get failed

File: src/util/IndexUtil.cpp [hunk 23]
```diff
@@ -22,6 +22,47 @@ Status IndexUtil::validateColumns(const std::vector<std::string>& fields) {
...
+        Row row;
+        row.values.emplace_back(Value(col.get_name()));
+        row.values.emplace_back(SchemaUtil::typeToString(col));
+        dataSet.emplace_back(std::move(row));
+    }
+    return dataSet;
+}
+
+StatusOr<DataSet> IndexUtil::toShowCreateIndex(bool isTagIndex,
+                                                const std::string &indexName,
```
Comment (laura-ding) [inline src/util/IndexUtil.cpp:37]: alignment

**Review by laura-ding — APPROVED** [review]

**Review by nevermore3 — APPROVED** [review]

**Review by yixinglu — APPROVED** [review]

**Pull request merged**